enhancements/storage/storage-performant-security-policy.md

```diff
@@ -0,0 +1,205 @@
+
+But for most intent and purpose, it can be a reasonable default and hence this proposal allows admins to configure a namespace such that, all pods created in the namespace will default to `OnRootMismatch` policy if pod does not specify a policy of its own. If a pod specifies its own `fsGroupChangePolicy` that is still used as a default and will not be overridden.
+
+### Motivation for seLinuxChangePolicy defaulting
+
+https://github.com/kubernetes/enhancements/issues/1710 is bringing changes which can be breaking for certain pods which use different selinux contexts for same volume between different pods or same pod. Once `SELinuxMount` feature becomes generally available, whenever supported (i.e CSI driver supports it, which will be true for all CSI drivers OCP ships by default), volumes will be mounted with selinux mount options, rather than container-runtime recursively `chcon`ing the volume.
+
+Applying SELinux label as a mount option by default will eliminate time spent by container-runtime in recursively changing selinux labels of a volume. We have determined this to be mostly safe by observing usage patterns and metrics collected during beta process. Having said that, we also are aware that certain workloads will be broken with this default. Workloads that use same volume in multiple pods or mount same volume in different containers with different selinux policies *may* break.
+
+By allowing cluster-admins to set namespace default of `seLinuxChangePolicy` we want to give control over how these changes are applied to Openshift clusters, without explicitly  changing pods themselves.
+
+### User Stories
+
+N/A
+
+### Goals
+
+- Allow Openshift admins to configure per-namespace default of `fsGroupChangePolicy`.
+- Allow Openshift admins to configure per-namespace default of `seLinuxChangePolicy`.
+
+### Non-Goals
+
+* Change Openshift default globally.
+
+## Proposal
+
+### Allow admins to opt-in to `fsGroupChangePolicy` via namespace label
+
+We propose a label `storage.openshift.io/fsgroup-change-policy` if that is set to `OnRootMismatch` in a namespace, then Openshift's SCC hooks start defaulting all pods created in that namespace with `OnRootMismatch` `FSGroupChangePolicy`.
+
+```go
+func getPodFsGroupChangePolicy(ns *corev1.Namespace) *api.PodFSGroupChangePolicy {
+    fsGroupPolicy, ok := ns.Labels[securityv1.FSGroupChangePolicy]
+    if !ok {
+        return nil
+    }
+    if fsGroupPolicy == "OnRootMismatch" {
+        onRootMismatchPolicy := api.FSGroupChangeOnRootMismatch
+        return &onRootMismatchPolicy
+    }
+    return nil
+}
+```
+
+### Allow admins to opt-in to `seLinuxChangePolicy` via namespace label
```

Comment:
Are there any restrictions on who is authorized to write either namespace label? Do there need to be any?

Reply:
I think this should fall in line with other namespace level activities that require cluster-admin role. A good example for example is adding following label for metric gathering:

```
    oc label namespace <namespace-name> openshift.io/cluster-monitoring="true"
```